Node doors

capital-G opened the pull request:
This contains #523 (but not everything, the new changes from you have to be cherry-picked) with an additional merge from main which was necessary because I needed the latest database migrations.

It introduces the concept of `NodeDoors` which replaces the old *naive* implementation of `Node --Edge--> Node`

Some screenshots from the docs

<img width="832" alt="grafik" src="https://github.com/Gencaster/gencaster/assets/8267062/e28711a2-1783-4860-9069-b5d193360fc9">

<img width="843" alt="grafik" src="https://github.com/Gencaster/gencaster/assets/8267062/dfc4f220-e239-4821-91f4-b6e1f10ba8be">

I have not tested it thoroughly in person but wrote some tests for the backend and the switch-logic should work.

Stuff that still needs to be done:

* [x] styling (maybe outside of this PR or in an additional branch?)
* [ ] documentation on how to use node doors within the editor  - with examples
* [x] ~dragging a connection is currently not possible - something broke it. The current way to create a connection is to first click on the target and afterwards on the source. Would be great to have some help here b/c I don't know what to do here :/ ~ => fixed in [961b765](https://github.com/Gencaster/gencaster/pull/553/commits/961b765d86648f37b7f26025cb74574ad74bc6d1
* [x] State management of node door code is wonky
  * Switching nodes with unapplied changes should not be possible
  * After switching we get the error "please apply changes" although the button is greyed out 
* [x] Check if python code is valid (see #540 )

Still a WIP, but maybe some feedback would be nice :)

capital-G commented:
@vin-ni I fixed most typescript issues except one

https://github.com/Gencaster/gencaster/blob/55f546ff49e4e876f5d22d88a6699ac22b9d1790/caster-editor/src/components/MenuTabEdit.vue#L76-L91

One would need to select both "Handles" of a node - I think we could drop the button "Create connection" in favor of drap&dropping connections - do you agree?

vin-ni commented:
> I think we could drop the button "Create connection" in favor of drap&dropping connections - do you agree?

Yes, defenitely

capital-G commented:
Deleted, now the CI should work w/o issues :)

vin-ni commented:
Added styling. 
There seems to be still a state bug. I can't add new scriptcells:

https://github.com/Gencaster/gencaster/assets/9806572/bbab372a-4ec8-4531-8818-843faa54b42d

capital-G commented:
Fixed two bugs
* Fixed creation of new script cells
* Fixed a bug where the deletion of a node door was not reflected in the node editor UI due to a race condition

For me this is ready to be merged? The docs could be added in another PR, what do you think @vin-ni ?

vin-ni commented on caster-back/story_graph/engine.py:
```diff
@@ -311,6 +323,79 @@ async def execute_node(
+        If multiple out-going edges are connected to an active door,
+        a random edge will be picked to follow for the next node.
+
+        If the node does not have any out-going edges a :class:`~GraphDeadEnd`
```
This should trigger the end-screen, right?

capital-G replied:
It should :tm:
But see  https://github.com/Gencaster/gencaster/issues/530#issuecomment-1712911247

vin-ni reviewed (approved):
Approved!
